### ISSUE TITLE
[BUG?] Float64 ends up being float32 if reading with externalsource + DALIGenericIterator

### ISSUE THREAD

**Issue by JuanFMontesinos** [NVIDIA/DALI#2191]
Hi,
I've realized that there is some sort of bug? while reading data with external source.

Context: I'm loading int16 audio using librosa. I mapped those values into -1,1 by dividing by the range our int16. Numpy's cast  promotion made those arrays to be float64. I was feeding and external source op via iter_setup. I was (as sanity check) retourning the same array using pytorch DALIGenericIterator and it was not bounded between -1 and 1. Besides it was float 32. 

In short,
It seems that DALI or DALIGenericIterator works with float32 and if you feed it float64  generates a silent error. 

Dunno if this was expected but it's really annoying

Sorry by the poor description.

**Comment by JanuszL** [comment]
Hi,
I'm afraid you need to be more specific. If you can provide a simple, self-contained reproduction it would help much.

I see that DALI propagates the types correctly. You can run this simple test on your own and check the result:
```
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types
import numpy as np
from nvidia.dali.plugin.pytorch import DALIGenericIterator

class TestPipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, num_gpus, test_array):
        super(TestPipeline, self).__init__(batch_size,
                                            num_threads,
                                            device_id)
        self.input_data = ops.ExternalSource()
        self.test_array = test_array

    def define_graph(self):
        self.data = self.input_data()
        return (self.data)


    def iter_setup(self):
        self.feed_input(self.data, self.test_array)

for t in [np.float64]:
    test_array = np.array([[1, 1], [1, 1]], dtype=t)
    pipe = TestPipeline(2, 1, 0, 1, test_array)
    pipe.build()
    # out = pipe.run()[0]
    # out_dtype = out.at(0).dtype
    # print(out.at(0).dtype)
    # print(test_array.dtype)
    dali_iter = DALIGenericIterator(pipe, ['data'], -1)
    for i, data in enumerate(dali_iter):
        print(data[0]["data"].dtype)
        print(test_array.dtype)
        break
```
In my case it prints:
```
torch.float64
float64
```

**Comment by JuanFMontesinos** [comment]
Hi,
First of all I'm impressed you could understand me with such a description :O 
Indeed your example is right. I also used ops.Spectrogram and that's what generates the issue.

Sorry for the annoying notation, as it's a complex library I just copied everything from my original code. 
```
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types
import numpy as np
from nvidia.dali.plugin.pytorch import DALIGenericIterator

class TestPipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, num_gpus, test_array):
        super(TestPipeline, self).__init__(batch_size,
                                            num_threads,
                                            device_id)
        self.audio_main_data = ops.ExternalSource()
        self.spectrogram = ops.Spectrogram(device="gpu",
                                           nfft=1022,
                                           window_length=1022,
                                           window_step=256,
                                           window_fn=[])
        self.test_array = [test_array for _ in range(2)]

    def define_graph(self):
        self.audio_main = self.audio_main_data().gpu()
        sp_main = self.spectrogram(self.audio_main)
        return sp_main,self.audio_main


    def iter_setup(self):
        self.feed_input(self.audio_main, self.test_array)

for t in [np.float64,np.float32]:
    print(f'TESTING {t}======')
    test_array = np.random.rand(65536).astype(t)
    pipe = TestPipeline(2, 1, 0, 1, test_array)
    pipe.build()
    dali_iter = DALIGenericIterator(pipe, ['sp','data'], -1)
    for i, data in enumerate(dali_iter):
        print(data[0]["data"].dtype)
        print(test_array.dtype)
        break
```

**Comment by JuanFMontesinos** [comment]
Oh and I should mention that passing float64 to Spectrogram generates wrong float32 spectrograms

**Comment by JanuszL** [comment]
I see the problem now. Apparently, DALI doesn't check the input type of the `Spectrogram` and just reinterprets it as float32.
We will update the docs and add a necessary check, in the mean time try to add a [`Cast` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.Cast) that would convert FP64 to FP32 data.

**Comment by JuanFMontesinos** [comment]
Hi, yep the check would be useful.  The main problem I found is this leads to a silent error (since the spectrogram is wrong too) and had to lose time finding the error.  

Thanks for your time
Regards
Juan

**Comment by JanuszL** [comment]
https://github.com/NVIDIA/DALI/pull/2192 should fix your problem.
